File: translations/android-java/app/src/main/java/com/apexkit/sdk/ApexKit.java
package com.apexkit.sdk;

import com.apexkit.sdk.models.*;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import okhttp3.*;
import java.io.IOException;
import java.lang.reflect.Type;
import java.util.HashMap;
import java.util.Map;
import java.util.List;
import java.util.concurrent.TimeUnit;

public class ApexKit {
    private String baseUrl;
    private String token;
    private User currentUser;
    private String scopeType = "root";
    private String scopeId = "";
    private OkHttpClient client;
    private Gson gson;

    public ApexKit(String baseUrl) {
        this.baseUrl = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;
        this.client = new OkHttpClient.Builder()
                .connectTimeout(10, TimeUnit.SECONDS)
                .readTimeout(10, TimeUnit.SECONDS)
                .build();
        this.gson = new Gson();
    }

    public void setToken(String token, User user) {
        this.token = token;
        if (user != null) {
            this.currentUser = user;
            setScopeFromTag(user.scope);
        }
    }

    private void setScopeFromTag(String tag) {
        if ("root".equals(tag)) {
            scopeType = "root";
            scopeId = "";
        } else if (tag != null && tag.startsWith("tenant:")) {
            scopeType = "tenant";
            scopeId = tag.substring(7);
        } else if (tag != null && tag.startsWith("sandbox:")) {
            scopeType = "sandbox";
            scopeId = tag.substring(8);
        }
    }

    private Response executeRequest(String method, String endpoint, Object body, Map<String, String> params, boolean isRoot) throws IOException, ApexError {
        String path = endpoint;
        if (!isRoot && !endpoint.startsWith("/api/v1")) {
            path = endpoint.startsWith("/") ? "/api/v1" + endpoint : "/api/v1/" + endpoint;
        }

        HttpUrl.Builder urlBuilder = HttpUrl.parse(baseUrl + path).newBuilder();
        if (params != null) {
            for (Map.Entry<String, String> entry : params.entrySet()) {
                urlBuilder.addQueryParameter(entry.getKey(), entry.getValue());
            }
        }

        Request.Builder requestBuilder = new Request.Builder().url(urlBuilder.build());
        if (token != null) {
            requestBuilder.header("Authorization", "Bearer " + token);
        }

        RequestBody requestBody = null;
        if (body != null) {
            requestBody = RequestBody.create(
                    MediaType.parse("application/json; charset=utf-8"),
                    gson.toJson(body)
            );
        } else if ("POST".equals(method) || "PUT".equals(method) || "PATCH".equals(method)) {
            requestBody = RequestBody.create(MediaType.parse("application/json; charset=utf-8"), "");
        }

        requestBuilder.method(method, requestBody);
        Response response = client.newCall(requestBuilder.build()).execute();

        if (!response.isSuccessful()) {
            String bodyStr = response.body().string();
            Map<String, Object> errorData;
            try {
                errorData = gson.fromJson(bodyStr, new TypeToken<Map<String, Object>>(){}.getType());
            } catch (Exception e) {
                errorData = new HashMap<>();
                errorData.put("message", bodyStr);
            }
            throw new ApexError(
                    (String) errorData.get("message"),
                    response.code(),
                    (String) errorData.get("error"),
                    errorData.get("details")
            );
        }

        return response;
    }

    public AuthNamespace auth() { return new AuthNamespace(); }
    public AdminsNamespace admins() { return new AdminsNamespace(); }
    public AiNamespace ai() { return new AiNamespace(); }
    public ScriptsNamespace scripts() { return new ScriptsNamespace(); }
    public FilesNamespace files() { return new FilesNamespace(); }

    public class AuthNamespace {
        public AuthResponse login(String email, String password) throws IOException, ApexError {
            Map<String, String> body = new HashMap<>();
            body.put("email", email);
            body.put("password", password);
            Response response = executeRequest("POST", "/auth/login", body, null, false);
            AuthResponse res = gson.fromJson(response.body().string(), AuthResponse.class);
            setToken(res.token, res.user);
            return res;
        }
        public AuthResponse register(String email, String password) throws IOException, ApexError {
            Map<String, String> body = new HashMap<>();
            body.put("email", email);
            body.put("password", password);
            Response response = executeRequest("POST", "/auth/register", body, null, false);
            AuthResponse res = gson.fromJson(response.body().string(), AuthResponse.class);
            setToken(res.token, res.user);
            return res;
        }
        public User getMe() throws IOException, ApexError {
            Response response = executeRequest("GET", "/auth/me", null, null, false);
            User user = gson.fromJson(response.body().string(), User.class);
            setScopeFromTag(user.scope);
            return user;
        }
        public void logout() {
            token = null;
            currentUser = null;
        }
    }

    public CollectionNamespace collection(String id) {
        return new CollectionNamespace(id);
    }

    public class CollectionNamespace {
        private String id;
        public CollectionNamespace(String id) { this.id = id; }

        public ListResult<BaseRecord> list(Map<String, String> options) throws IOException, ApexError {
            Response response = executeRequest("GET", "/collections/" + id + "/records", null, options, false);
            Type type = new TypeToken<ListResult<BaseRecord>>(){}.getType();
            return gson.fromJson(response.body().string(), type);
        }
        public BaseRecord create(Map<String, Object> data) throws IOException, ApexError {
            Map<String, Object> body = new HashMap<>();
            body.put("data", data);
            Response response = executeRequest("POST", "/collections/" + id + "/records", body, null, false);
            return gson.fromJson(response.body().string(), BaseRecord.class);
        }
        public BaseRecord get(String recordId, Map<String, String> options) throws IOException, ApexError {
            Response response = executeRequest("GET", "/collections/" + id + "/records/" + recordId, null, options, false);
            return gson.fromJson(response.body().string(), BaseRecord.class);
        }
        public BaseRecord update(String recordId, Map<String, Object> data) throws IOException, ApexError {
            Map<String, Object> body = new HashMap<>();
            body.put("data", data);
            Response response = executeRequest("PUT", "/collections/" + id + "/records/" + recordId, body, null, false);
            return gson.fromJson(response.body().string(), BaseRecord.class);
        }
        public void delete(String recordId) throws IOException, ApexError {
            executeRequest("DELETE", "/collections/" + id + "/records/" + recordId, null, null, false);
        }
    }

    public class AdminsNamespace {
        public List<Map<String, Object>> listCollections() throws IOException, ApexError {
            Response response = executeRequest("GET", "/collections", null, null, false);
            return gson.fromJson(response.body().string(), new TypeToken<List<Map<String, Object>>>(){}.getType());
        }
    }

    public class AiNamespace {
        public Map<String, Object> run(String slug, Map<String, Object> variables) throws IOException, ApexError {
            Map<String, Object> body = new HashMap<>();
            body.put("variables", variables);
            Response response = executeRequest("POST", "/ai/run/" + slug, body, null, false);
            return gson.fromJson(response.body().string(), new TypeToken<Map<String, Object>>(){}.getType());
        }
    }

    public class ScriptsNamespace {
        public Object run(String name, Object variables) throws IOException, ApexError {
            Response response = executeRequest("POST", "/run/" + name, variables, null, false);
            return gson.fromJson(response.body().string(), Object.class);
        }
    }

    public class FilesNamespace {
        public String getFileUrl(String filename) {
            if (filename.startsWith("http")) return filename;
            return baseUrl + "/api/v1/storage/file/" + (filename.startsWith("/") ? filename.substring(1) : filename);
        }
    }

    public Map<String, Object> graphql(String query, Map<String, Object> variables) throws IOException, ApexError {
        Map<String, Object> body = new HashMap<>();
        body.put("query", query);
        body.put("variables", variables != null ? variables : new HashMap<>());
        Response response = executeRequest("POST", "/graphql", body, null, true);
        return gson.fromJson(response.body().string(), new TypeToken<Map<String, Object>>(){}.getType());
    }
}

File: translations/android-kotlin/app/src/main/kotlin/com/apexkit/sdk/ApexKit.kt
package com.apexkit.sdk

import com.apexkit.sdk.models.*
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

class ApexKit(baseUrl: String) {
    private val baseUrl = baseUrl.removeSuffix("/")
    private var token: String? = null
    private var currentUser: User? = null
    private var scopeType: String = "root"
    private var scopeId: String = ""
    private val client = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .build()
    private val gson = Gson()

    fun sandbox(uuid: String): ApexKit {
        val instance = ApexKit("${this.baseUrl}/sandbox/$uuid")
        instance.scopeType = "sandbox"
        instance.scopeId = uuid
        instance.setToken(this.token ?: "", this.currentUser)
        return instance
    }

    fun tenant(tenantId: String): ApexKit {
        val instance = ApexKit("${this.baseUrl}/tenant/$tenantId")
        instance.scopeType = "tenant"
        instance.scopeId = tenantId
        instance.setToken(this.token ?: "", this.currentUser)
        return instance
    }

    fun setToken(token: String, user: User?) {
        this.token = token
        if (user != null) {
            this.currentUser = user
            setScopeFromTag(user.scope)
        }
    }

    private fun setScopeFromTag(tag: String) {
        when {
            tag == "root" -> {
                scopeType = "root"
                scopeId = ""
            }
            tag.startsWith("tenant:") -> {
                scopeType = "tenant"
                scopeId = tag.removePrefix("tenant:")
            }
            tag.startsWith("sandbox:") -> {
                scopeType = "sandbox"
                scopeId = tag.removePrefix("sandbox:")
            }
        }
    }

    private fun request(
        method: String,
        endpoint: String,
        body: Any? = null,
        params: Map<String, String>? = null,
        isRoot: Boolean = false
    ): Response {
        var path = endpoint
        if (!isRoot && !endpoint.startsWith("/api/v1")) {
            path = if (endpoint.startsWith("/")) "/api/v1$endpoint" else "/api/v1/$endpoint"
        }

        val urlBuilder = HttpUrl.parse("$baseUrl$path")!!.newBuilder()
        params?.forEach { (k, v) -> urlBuilder.addQueryParameter(k, v) }

        val requestBuilder = Request.Builder().url(urlBuilder.build())
        token?.let { requestBuilder.header("Authorization", "Bearer $it") }

        val requestBody = if (body != null) {
            val mediaType = "application/json; charset=utf-8".toMediaType()
            gson.toJson(body).toRequestBody(mediaType)
        } else if (method == "POST" || method == "PUT" || method == "PATCH") {
            "".toRequestBody(null)
        } else {
            null
        }

        requestBuilder.method(method, requestBody)
        val response = client.newCall(requestBuilder.build()).execute()

        if (!response.isSuccessful) {
            val bodyStr = response.body?.string() ?: ""
            val errorData = try {
                gson.fromJson(bodyStr, Map::class.java)
            } catch (e: Exception) {
                mapOf("message" to bodyStr)
            }
            throw ApexError(
                errorData["message"] as? String ?: "API Error",
                response.code,
                errorData["error"] as? String,
                errorData["details"]
            )
        }

        return response
    }

    val auth = AuthNamespace()
    val admins = AdminsNamespace()
    val ai = AiNamespace()
    val scripts = ScriptsNamespace()
    val files = FilesNamespace()

    inner class AuthNamespace {
        fun login(email: String, password: String): AuthResponse {
            val response = request("POST", "/auth/login", mapOf("email" to email, "password" to password))
            val authRes = gson.fromJson(response.body?.string(), AuthResponse::class.java)
            setToken(authRes.token, authRes.user)
            return authRes
        }

        fun register(email: String, password: String): AuthResponse {
            val response = request("POST", "/auth/register", mapOf("email" to email, "password" to password))
            val authRes = gson.fromJson(response.body?.string(), AuthResponse::class.java)
            setToken(authRes.token, authRes.user)
            return authRes
        }

        fun getMe(): User {
            val response = request("GET", "/auth/me")
            val user = gson.fromJson(response.body?.string(), User::class.java)
            setScopeFromTag(user.scope)
            return user
        }

        fun logout() {
            token = null
            currentUser = null
        }
    }

    fun collection(id: String) = CollectionNamespace(id)

    inner class CollectionNamespace(val id: String) {
        fun list(options: Map<String, String>? = null): ListResult<BaseRecord> {
            val response = request("GET", "/collections/$id/records", params = options)
            val type = object : TypeToken<ListResult<BaseRecord>>() {}.type
            return gson.fromJson(response.body?.string(), type)
        }

        fun create(data: Map<String, Any>): BaseRecord {
            val response = request("POST", "/collections/$id/records", body = mapOf("data" to data))
            return gson.fromJson(response.body?.string(), BaseRecord::class.java)
        }

        fun get(recordId: String, options: Map<String, String>? = null): BaseRecord {
            val response = request("GET", "/collections/$id/records/$recordId", params = options)
            return gson.fromJson(response.body?.string(), BaseRecord::class.java)
        }

        fun update(recordId: String, data: Map<String, Any>): BaseRecord {
            val response = request("PUT", "/collections/$id/records/$recordId", body = mapOf("data" to data))
            return gson.fromJson(response.body?.string(), BaseRecord::class.java)
        }

        fun patch(recordId: String, data: Map<String, Any>): BaseRecord {
            val response = request("PATCH", "/collections/$id/records/$recordId", body = mapOf("data" to data))
            return gson.fromJson(response.body?.string(), BaseRecord::class.java)
        }

        fun delete(recordId: String) {
            request("DELETE", "/collections/$id/records/$recordId")
        }
    }

    inner class AdminsNamespace {
        fun listCollections(): List<Map<String, Any>> {
            val response = request("GET", "/collections")
            val type = object : TypeToken<List<Map<String, Any>>>() {}.type
            return gson.fromJson(response.body?.string(), type)
        }
    }

    inner class AiNamespace {
        fun run(slug: String, variables: Map<String, Any>): Map<String, Any> {
            val response = request("POST", "/ai/run/$slug", mapOf("variables" to variables))
            val type = object : TypeToken<Map<String, Any>>() {}.type
            return gson.fromJson(response.body?.string(), type)
        }
    }

    inner class ScriptsNamespace {
        fun run(name: String, variables: Any): Any {
            val response = request("POST", "/run/$name", variables)
            val type = object : TypeToken<Any>() {}.type
            return gson.fromJson(response.body?.string(), type)
        }
    }

    inner class FilesNamespace {
        fun getFileUrl(filename: String): String {
            if (filename.startsWith("http")) return filename
            return "$baseUrl/api/v1/storage/file/${filename.removePrefix("/")}"
        }
    }

    fun graphql(query: String, variables: Map<String, Any>? = null): Map<String, Any> {
        val response = request("POST", "/graphql", mapOf("query" to query, "variables" to (variables ?: emptyMap<String, Any>())), isRoot = true)
        val type = object : TypeToken<Map<String, Any>>() {}.type
        return gson.fromJson(response.body?.string(), type)
    }
}
